File: 1129-longest-string-chain/longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Arrays.sort(words,(a,b)->a.length()-b.length());
        HashMap<String,Integer> mp=new HashMap<>();
        int ans=1;
        for(int i=0;i<words.length;i++){
            String s=words[i];
            mp.put(s,1);
            for(int j=0;j<s.length();j++){
                String pre=s.substring(0,j)+s.substring(j+1);
                if(mp.containsKey(pre)){
                    mp.put(s,Math.max(mp.get(s), mp.get(pre) + 1));
                }
            }
            ans=Math.max(ans,mp.get(s));
        }
        return ans;
    }
}

File: 1129-longest-string-chain/longest-string-chain.cpp
class Solution {
public:
    static bool compare(string a,string b){
        return a.length()<b.length();
    }
    int longestStrChain(vector<string>& words) {
        int ans=0;
        sort(words.begin(),words.end(),compare);
        unordered_map<string,int> mp;
        for(string t:words){
            mp[t]=1;
            for(int i=0;i<t.length();i++){
                string s=t;
                s.erase(i,1);
                if(mp.count(s) && mp[s]+1>mp[t]){
                    mp[t]=mp[s]+1;
                }
            }
            ans=max(ans,mp[t]);
        }
        return ans;
    }
};